### ISSUE TITLE
lazy-ignores: a paraphrased Suppressions header entry is treated as unjustified

### ISSUE THREAD

**Issue by be-wise-be-kind** [be-wise-be-kind/thai-lint#249]
## Summary

`lazy-ignores` reports a suppression as "unjustified" (and its matching `Suppressions:` header
entry as "orphaned") when the inline comment's reason text is a *paraphrase* of the header
explanation rather than a near-literal echo of it — even though the header entry genuinely,
unambiguously justifies the same rule code on the same finding. Matching a genuine but differently
worded justification is exactly the case this check exists to allow; a check that only accepts
near-identical wording defeats the purpose of writing the header prose in the reviewer's own words.

## Environment

- `thailint`, version 0.22.1 (via `uvx thailint@latest`)
- Reproduced with a single file, in a fresh, empty git repo — no other project config involved.

## Minimal repro

`example.py` (fails):

```python
"""
File: example.py

Purpose: Minimal repro for a lazy-ignores false positive.

Suppressions:
    - S607: `git` is spelled without an absolute path; the argv is otherwise literal and the
      call only reads the tracked-file list.
"""

import subprocess


def run():
    return subprocess.run(["git", "ls-files"], capture_output=True, text=True)  # noqa: S607 — `git` off PATH, as elsewhere here
```

```
$ uvx thailint@latest lazy-ignores example.py
Found 1 violation(s):

  example.py:15:81
    [ERROR] lazy-ignores.unjustified: Unjustified suppression found: # noqa: S607 — `git` off PATH, as elsewhere here (ASK PERMISSION before adding Suppressions header)
```

`example_control.py` (passes) — same finding, same rule, same underlying justification, wording
echoed instead of paraphrased:

```python
"""
File: example_control.py

Purpose: Control case for the lazy-ignores repro — same finding, wording echoed instead of
    paraphrased.

Suppressions:
    - S607: literal argv, no shell, git off PATH.
"""

import subprocess


def run():
    return subprocess.run(["git", "ls-files"], capture_output=True, text=True)  # noqa: S607 — literal argv, no shell, git off PATH
```

```
$ uvx thailint@latest lazy-ignores example_control.py
✓ No violations found
```

The only difference between the two files is the wording of the header bullet and the inline
comment — same rule code (`S607`), same code shape, same semantic justification ("git is invoked
without an absolute path, but the argv is otherwise fixed/literal").

## Impact on a real project

This surfaced on a real repo (not included here, since it's private) where several suppressions
across multiple files failed the same way despite each having a clearly matching, human-written
header entry — including entries reusing an *established* explanation by cross-reference (e.g. "same
idiom as file X's own Suppressions entry") rather than restating it, and multi-code headers where a
bullet for one rule code appears to interfere with matching a different bullet for a different rule
code on the same file. I can only speak confidently to the paraphrase-vs-echo shape shown above,
since it's the one I isolated to a clean, deterministic, single-file repro — the multi-file/
cross-reference cases may be the same root cause or a separate one; flagging them here as
unconfirmed context in case they're useful in tracking down the root cause.

## Expected behavior

A `Suppressions:` header entry that unambiguously names the same rule code and describes the same
finding as an inline suppression comment should satisfy `lazy-ignores`, regardless of whether the
inline comment's short reason phrase is a verbatim substring of the header prose or an independent
paraphrase of the same justification.

**Comment by be-wise-be-kind** [comment]
Thank you for the exceptionally clear report — the minimal repro plus the control file made the root cause easy to isolate. We're investigating now and will follow up shortly.

**Comment by be-wise-be-kind** [comment]
## Follow-up: isolated the actual trigger

Not wording similarity after all — it's whether the **first token of the bullet's body** (right
after `CODE: `) is wrapped in backticks. Same header prose, same inline comment, word-for-word:

```python
"""
Suppressions:
    - S607: `git` is spelled without an absolute path; the argv is otherwise literal.
"""
...
    return subprocess.run(["git", "ls-files"], ...)  # noqa: S607 — git is spelled without an absolute path; the argv is otherwise literal
```
```
$ uvx thailint@latest lazy-ignores example2.py
Found 1 violation(s):
  [ERROR] lazy-ignores.unjustified: ...
```

Remove only the backticks around the header bullet's opening word (`` `git` `` → `git`), nothing
else:

```python
"""
Suppressions:
    - S607: git is spelled without an absolute path; the argv is otherwise literal.
"""
```
```
$ uvx thailint@latest lazy-ignores example2_fixed.py
✓ No violations found
```

Reproduced the same way on a real, larger file (a `PLR0913` bullet starting `` `_below_threshold_stats` takes 6 arguments... ``) — removing the backtick-quoted, underscore-prefixed
function name from the start of that bullet fixed it too, with the rest of the sentence
unchanged. So it isn't specific to backticks-around-`git` or to leading underscores individually;
it looks like the parser is tripping on the first bullet-body token being wrapped in backticks in
general (I'd guess it's read as a code-span and something downstream doesn't handle that at the
very start of the sentence, e.g. an index calculation that assumes the first token is plain text).
This is likely the same root cause behind the whole original report, and a much sharper
description of it than the "paraphrase vs. echo" framing in the description above — that
framing was a coincidence of which examples I'd tried, not the real variable.

Workaround for anyone hitting this in the meantime: don't start a Suppressions bullet's body with
a backtick-quoted term — rephrase so the sentence opens with plain text instead.

**Comment by be-wise-be-kind** [comment]
Root cause found, and it is not paraphrase-vs-echo — so here is what is actually happening plus two workarounds you can apply today, without waiting for the release.

## What is actually broken

The header entry in your repro is **never parsed at all**. The `Suppressions:` entry regex requires the justification text to begin with a letter or underscore. Yours begins with a backtick:

```
- S607: `git` is spelled without an absolute path; ...
   ^ justification starts with ` -> entry silently dropped
```

With the entry dropped, `S607` has no header justification, so the inline comment is reported unjustified. Your control file passes only because `literal argv, ...` happens to start with a letter. Rewording it to a paraphrase that also started with a letter would have passed too.

Two related defects in the same regex, which we believe cover the unconfirmed cases you flagged:

1. **Greedy split on the last colon.** `- S607: note: git is off PATH` parses as rule id `s607: note`, so the real `S607` goes unmatched. Any justification containing `": "` hits this.
2. **Wrapped continuation lines are parsed as new entries.** A second line of a wrapped justification that contains `": "` becomes a bogus entry — which then reports as *orphaned*, and can shadow a sibling bullet for a different rule code. That matches your multi-code and orphaned observations.

Anything that starts a justification with a non-letter is affected: backticks, quotes, digits, parentheses.

## Workarounds (both verified against the repro)

**A. Reword the header entry** so the justification starts with a letter, stays on one line, and contains no `": "`:

```python
"""
Suppressions:
    - S607: git is spelled without an absolute path; the argv is otherwise literal and the call only reads the tracked-file list.
"""
```

**B. Or skip the header and justify inline** with a space-hyphen-space separator and at least 10 characters of reason. This bypasses header matching entirely:

```python
subprocess.run(["git", "ls-files"], ...)  # noqa: S607 - `git` off PATH, argv is literal
```

Note that B needs an ASCII `-`. The em dash `—` in your repro is not recognized as an inline separator today — that is a separate gap, and we are fixing it in the same release.

## The fix

We are replacing the line-level regex with structure-aware entry parsing: bullet/indent grouping so wrapped lines stay with their entry, splitting on the first colon, and accepting any justification text regardless of its first character. Em and en dashes will also be accepted as inline separators.

Thanks again — the control file is what made this a 10-minute diagnosis instead of a 2-hour one.

**Comment by be-wise-be-kind** [comment]
Fixed in **thailint v0.22.2**, now on PyPI.

```bash
pip install --upgrade thailint
# or: uvx thailint@latest lazy-ignores example.py
```

Your `example.py` repro now reports clean, verified against a fresh install of 0.22.2 from PyPI.

## What was wrong

Three defects in the `Suppressions:` entry regex, all fixed:

| Header entry | Before | After |
|---|---|---|
| ``- S607: `git` is spelled without…`` | entry dropped, so the noqa read as unjustified | matches `S607` |
| `- PLR0912: 12 branches are inherent…` | entry dropped (leading digit) | matches `PLR0912` |
| `- S607: note: git is off PATH` | keyed on `s607: note`, real rule unmatched | matches `S607` |
| a wrapped line containing `": "` | became a phantom entry, reported orphaned | joined to its entry |

The first row is your repro. The justification had to begin with a letter or underscore, so a leading backtick dropped the entry before matching ever ran. Paraphrasing was not the variable — your control file passed because `literal argv, …` happens to start with a letter.

The last two rows are your unconfirmed cases. Your instinct was right: a phantom entry from a wrapped line is exactly what produces a spurious orphan, and in a multi-entry section it could shadow a sibling bullet for a different rule code. The cross-reference case ("same idiom as file X's entry") should also be resolved, since the entry now parses whatever its prose says — if you still see that one, please reopen with a repro.

The entry regex is replaced by a structure-aware parser: bullet and indentation cues group continuation lines with their entry, the rule ID ends at the first colon followed by whitespace, and justification text may start with any character.

## Also in this release

Em dash and en dash now work as inline justification separators, so the inline form in your repro is valid on its own:

```python
subprocess.run(["git", "ls-files"], ...)  # noqa: S607 — `git` off PATH, argv is literal
```

`docs/lazy-ignores-linter.md` documents the entry format rules and inline justifications, which had never been written up.

Thank you again for the report. The control file is what turned this into a fast diagnosis, and it directly surfaced two defects beyond the one you isolated.
